subsys/usb/device_next/usbd_core.c

```diff
@@ -25,16 +25,26 @@
 #include <zephyr/logging/log.h>
 LOG_MODULE_REGISTER(usbd_core, CONFIG_USBD_LOG_LEVEL);
 
-static K_KERNEL_STACK_DEFINE(usbd_stack, CONFIG_USBD_THREAD_STACK_SIZE);
-static struct k_thread usbd_thread_data;
-
 K_MSGQ_DEFINE(usbd_msgq, sizeof(struct udc_event),
 	      CONFIG_USBD_MAX_UDC_MSG, sizeof(uint32_t));
 
+static void usbd_flush_queue(void);
+
 static int usbd_event_carrier(const struct device *dev,
 			      const struct udc_event *const event)
 {
-	return k_msgq_put(&usbd_msgq, event, K_NO_WAIT);
+	int ret;
+
+	LOG_DBG("Submitting event %p from %s", event, dev->name);
+
+	ret = k_msgq_put(&usbd_msgq, event, K_NO_WAIT);
+	if (ret != 0) {
+		return ret;
+	}
+
+	usbd_flush_queue();
```

Comment:
No, thanks. We are not going to execute anything in the ISR context and continue the nightmare of the legacy stack. It is true that for some simple applications the request callbacks could be called without queuing, but the order of the events will be lost and may cause some problems, and the events that are still submitted from the ISR context will have to be offloaded to a thread. Overall, there is not much to gain from this. Also, k_msgq_put/usbd_flush_queue combo does not make any sense to me.

Reply:
> We are not going to execute anything in the ISR context and continue the nightmare of the legacy stack

An USB stack that works is better than an USB stack that fails fast. :sweat_smile: